File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.cs
using System;
using System.Collections.Generic;

public class Solution {
    public int[] solution(string today, string[] terms, string[] privacies) {
        var db = new Dictionary<string, int>();
        
        foreach(var t in terms)
        {
            var parts = t.Split(' ');
            db[parts[0]] = int.Parse(parts[1]);
        }
        
        int totalDays = ChangeDays(today);
        
        var answer = new List<int>();
        for(int i = 0; i < privacies.Length; i++)
        {
            var parts = privacies[i].Split(' ');
            string days = parts[0];
            string term = parts[1];
            
            int total = ChangeDays(days);
            int month = db[parts[1]];
            
            int expireDay = total + month * 28 - 1;
            
            if(expireDay < totalDays)
            {
                answer.Add(i + 1);
            }
        }
        answer.Sort();
        
        return answer.ToArray();
    }
    
    private int ChangeDays(string yyyymmdd)
    {
        var s = yyyymmdd.Split('.');
        int y = int.Parse(s[0]);
        int m = int.Parse(s[1]);
        int d = int.Parse(s[2]);
        
        return ((y * 12) + (m - 1)) * 28 + d;
    }
}

File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.cpp
#include <string>
#include <vector>
#include <bits/stdc++.h>

using namespace std;

int change(string date)
{
    int y = stoi(date.substr(0, 4));
    int m = stoi(date.substr(5, 2));
    int d = stoi(date.substr(8, 2));
    
    return y * 12 * 28 + m * 28 + d;
}
vector<int> solution(string today, vector<string> terms, vector<string> privacies) {
    vector<int> answer;
    
    unordered_map<string, int> mp;
    
    for(string term : terms)
    {
        string type = term.substr(0, 1);
        int month = stoi(term.substr(2));
        
        mp[type] = month;
    }
    
    int todayNum = change(today);
    
    for(int i = 0; i < privacies.size(); i++)
    {
        string privacy = privacies[i];
        
        string date = privacy.substr(0, 10);
        string type = privacy.substr(11, 1);
        
        int expire = change(date) + mp[type] * 28;
        
        if(expire <= todayNum)
            answer.push_back(i + 1);
    }
    return answer;
}